### PR TITLE
BIDS-2229/totalSyncAttestationStats

### PR DESCRIPTION
**Pull request by LuccaBitfly**
Adds `missed_attestations_total`, `participated_sync_total`, `missed_sync_total` and `orphaned_sync_total` columns to the `validator_stats` table

### <samp>🤖 Generated by Copilot at 7cf3a5e</samp>

This pull request enhances the validator performance metrics by adding new columns to the `validator_stats` table and the `Exported` struct. It also modifies the `db/statistics.go` file to use the new data in the queries and error handling. The changes are implemented using a migration file `db/migrations/tbd_add_total_attestations_and_sync.sql` that alters the table schema.

### CODE REVIEW

File: db/statistics.go [hunk 1]
```diff
@@ -181,18 +181,20 @@ func WriteValidatorTotalPerformance(day uint64) error {
...
 	type Exported struct {
-		LastClRewards    bool `db:"last_cl_rewards_exported"`
-		LastElRewards    bool `db:"last_el_rewards_exported"`
-		CurrentCLRewards bool `db:"cur_cl_rewards_exported"`
-		CurrentElRewards bool `db:"cur_el_rewards_exported"`
+		LastTotalPerformance      bool `db:"last_total_performance_exported"`
+		CurrentCLRewards          bool `db:"cur_cl_rewards_exported"`
+		CurrentElRewards          bool `db:"cur_el_rewards_exported"`
+		CurrentSyncDuties         bool `db:"cur_el_rewards_exported"`
+		CurrentFailedAttestations bool `db:"cur_el_rewards_exported"`
```
Comment (D13ce) [inline db/statistics.go:188]: Copy&past error for `db` tag.

File: db/migrations/tbd_add_total_attestations_and_sync.sql
Comment (D13ce) [inline db/migrations/tbd_add_total_attestations_and_sync.sql:1]: Reminder comment so we don't forget to rename this before merging the PR.

File: db/statistics.go [hunk 2]
```diff
@@ -201,8 +203,9 @@ func WriteValidatorTotalPerformance(day uint64) error {
 
 	if err != nil {
 		return fmt.Errorf("error retrieving required data: %v", err)
-	} else if !exported.CurrentCLRewards || !exported.CurrentElRewards || !exported.LastClRewards || !exported.LastElRewards {
-		return fmt.Errorf("missing required export: cur cl rewards: %v, cur el rewards: %v, last cl rewards: %v, last el rewards: %v", !exported.CurrentCLRewards, !exported.CurrentElRewards, !exported.LastClRewards, !exported.LastElRewards)
+	} else if !exported.LastTotalPerformance || !exported.CurrentCLRewards || !exported.CurrentElRewards || !exported.CurrentSyncDuties || !exported.CurrentFailedAttestations {
```
Comment (D13ce) [inline db/statistics.go:206]: As discussed, this check will fail for day 0 as you now have the following logic:
Export Total Performance for day 0 only if Total Performance for day 0 is exported. Mind the `cur.day - 1` here: https://github.com/gobitfly/eth2-beaconchain-explorer/blob/7cf3a5e2ebb03ab20ba5dc1d01981f2765d62533/db/statistics.go#L200

**Review by D13ce — CHANGES_REQUESTED** [review]

File: db/statistics.go (hunk 1 above)
Comment (LuccaBitfly) [reply db/statistics.go]: fixed

File: db/statistics.go (hunk 2 above)
Comment (LuccaBitfly) [reply db/statistics.go]: fixed

**Review by D13ce — APPROVED** [review]
utACK
Last thing left to do right before merging the PR: https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2354#discussion_r1242204099

File: db/statistics.go [hunk 3]
```diff
@@ -181,18 +181,20 @@ func WriteValidatorTotalPerformance(day uint64) error {
...
 	type Exported struct {
-		LastClRewards    bool `db:"last_cl_rewards_exported"`
-		LastElRewards    bool `db:"last_el_rewards_exported"`
-		CurrentCLRewards bool `db:"cur_cl_rewards_exported"`
-		CurrentElRewards bool `db:"cur_el_rewards_exported"`
+		LastTotalPerformance      bool `db:"last_total_performance_exported"`
+		CurrentCLRewards          bool `db:"cur_cl_rewards_exported"`
+		CurrentElRewards          bool `db:"cur_el_rewards_exported"`
+		CurrentSyncDuties         bool `db:"cur_sync_duties_exported"`
+		CurrentFailedAttestations bool `db:"cur_failed_attestations_exported"`
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply db/statistics.go:188]: They were created during BIDS-2171 (see this [PR](https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2263)).

File: db/statistics.go [hunk 4]
```diff
@@ -227,21 +230,43 @@ func WriteValidatorTotalPerformance(day uint64) error {
...
 						vs1.day, 
 						COALESCE(vs1.cl_rewards_gwei, 0) + COALESCE(vs2.cl_rewards_gwei_total, 0) AS cl_rewards_gwei_total_new, 
 						COALESCE(vs1.cl_proposer_rewards_gwei, 0) + COALESCE(vs2.cl_proposer_rewards_gwei_total, 0) AS cl_proposer_rewards_gwei_total_new, 
 						COALESCE(vs1.el_rewards_wei, 0) + COALESCE(vs2.el_rewards_wei_total, 0) AS el_rewards_wei_total_new, 
-						COALESCE(vs1.mev_rewards_wei, 0) + COALESCE(vs2.mev_rewards_wei_total, 0) AS mev_rewards_wei_total_new 
+						COALESCE(vs1.mev_rewards_wei, 0) + COALESCE(vs2.mev_rewards_wei_total, 0) AS mev_rewards_wei_total_new,
+						COALESCE(vs1.missed_attestations, 0) + COALESCE(vs2.missed_attestations_total, 0) AS missed_attestations_total_new,
+						COALESCE(vs1.participated_sync, 0) + COALESCE(vs2.participated_sync_total, 0) AS participated_sync_total_new,
+						COALESCE(vs1.missed_sync, 0) + COALESCE(vs2.missed_sync_total, 0) AS missed_sync_total_new,
+						COALESCE(vs1.orphaned_sync, 0) + COALESCE(vs2.orphaned_sync_total, 0) AS orphaned_sync_total_new
```
[the remark this answers is not in the document]
Comment (LuccaBitfly) [reply db/statistics.go]: probably for testing purposes, removed them, ptal

File: db/statistics.go [hunk 5]
```diff
@@ -248,14 +248,14 @@ func WriteValidatorTotalPerformance(day uint64) error {
...
 						vs1.day, 
-						COALESCE(vs1.cl_rewards_gwei, 0) + COALESCE(vs2.cl_rewards_gwei_total, 0) AS cl_rewards_gwei_total_new, 
-						COALESCE(vs1.cl_proposer_rewards_gwei, 0) + COALESCE(vs2.cl_proposer_rewards_gwei_total, 0) AS cl_proposer_rewards_gwei_total_new, 
-						COALESCE(vs1.el_rewards_wei, 0) + COALESCE(vs2.el_rewards_wei_total, 0) AS el_rewards_wei_total_new, 
-						COALESCE(vs1.mev_rewards_wei, 0) + COALESCE(vs2.mev_rewards_wei_total, 0) AS mev_rewards_wei_total_new,
-						COALESCE(vs1.missed_attestations, 0) + COALESCE(vs2.missed_attestations_total, 0) AS missed_attestations_total_new,
-						COALESCE(vs1.participated_sync, 0) + COALESCE(vs2.participated_sync_total, 0) AS participated_sync_total_new,
-						COALESCE(vs1.missed_sync, 0) + COALESCE(vs2.missed_sync_total, 0) AS missed_sync_total_new,
-						COALESCE(vs1.orphaned_sync, 0) + COALESCE(vs2.orphaned_sync_total, 0) AS orphaned_sync_total_new
+						COALESCE(vs1.cl_rewards_gwei, 0) + COALESCE(vs2.cl_rewards_gwei_total, 0) AS,
```
Comment (peterbitfly) [inline db/statistics.go:251]: please remove the `AS` as it will cause a syntax error

Comment (LuccaBitfly) [reply db/statistics.go]: fixed, ptal

**Comment by LuccaBitfly** [comment]
removed export script again

**Pull request merged**